python/hsfs/core/training_dataset_engine.py

```diff
@@ -107,3 +113,81 @@ def delete_tag(self, training_dataset, name):
     def get_tags(self, training_dataset, name):
         """Get tag with a certain name or all tags for a training dataset."""
         return [tag.to_dict() for tag in self._tags_api.get(training_dataset, name)]
+
+    def get_serving_vector(self, training_dataset, entry):
+        """Get ... for a training dataset."""
+
+        serving_vector = []
+
+        if training_dataset.prepared_statements is None:
+            self._init_prepared_statement(training_dataset)
+
+        prepared_statements = training_dataset.prepared_statements
+
+        for prepared_statement_index in prepared_statements:
+            prepared_statement = prepared_statements[prepared_statement_index]
+            # TODO (davit) check if this returns anything before fetchall
+            result_proxy = training_dataset.prepared_statement_connection.execute(
+                prepared_statement, entry
+            ).fetchall()
+            for row in result_proxy:
+                result_dict = dict(row.items())
+            serving_vector += list(result_dict.values())
+
+        return serving_vector
+
+    def _init_prepared_statement(self, training_dataset):
+        online_conn = self._storage_connector_api.get_online_connector()
+        jdbc_connection = self._create_mysql_connection(online_conn)
+        prepared_statements = self._training_dataset_api.get_serving_prepared_statement(
+            training_dataset
+        )
+
+        prepared_statements_dict = {}
+        for prepared_statement in prepared_statements:
+            query_online = str(prepared_statement.query_online).replace("\n", "")
+
+            # In java prepared statement `?` is used for prepared statement parametrization.
+            # In sqlalchemy `:feature_name` is used instead of `?`
+            pk_names = [
+                psp.name
+                for psp in sorted(
+                    prepared_statement.prepared_statement_parameters,
+                    key=lambda psp: psp.index,
+                )
+            ]
+            for pk_name in pk_names:
+                query_online = re.sub(
+                    r"^(.*?(\?.*?){0})\?",
+                    r"\1:" + pk_name,
+                    query_online,
```

Comment:
there is no other way to do this that's more readable than a regex?

Reply:
this was the shortest. May be I can annotate what regex is doing? Otherwise it will be separate function with couple of lines and loops

Reply:
yeah add some comments at least then